File: src/main/java/trivia/Game.py
from collections import deque

class Player:
    def __init__(self, name):
        self.name = name
        self.position = 1
        self.coins = 0
        self.in_penalty_box = False
        self.is_getting_out = False

    def add_coin(self):
        self.coins += 1

    def send_to_penalty_box(self):
        self.in_penalty_box = True

    def has_won(self, winning_coins):
        return self.coins == winning_coins

    def advance(self, roll, board_size):
        self.position += roll
        if self.position > board_size:
            self.position -= 12

class QuestionDeck:
    def __init__(self):
        self.pop_questions = deque([f"Pop Question {i}" for i in range(50)])
        self.science_questions = deque([f"Science Question {i}" for i in range(50)])
        self.sports_questions = deque([f"Sports Question {i}" for i in range(50)])
        self.rock_questions = deque([f"Rock Question {i}" for i in range(50)])

    def draw_question(self, category):
        if category == "Pop": return self.pop_questions.popleft()
        if category == "Science": return self.science_questions.popleft()
        if category == "Sports": return self.sports_questions.popleft()
        if category == "Rock": return self.rock_questions.popleft()
        return "Unknown Category"

class Game:
    BOARD_SIZE = 12
    WINNING_COINS = 6
    MAX_PLAYERS = 6

    def __init__(self):
        self.players = []
        self.deck = QuestionDeck()
        self.current_player_idx = 0

    @property
    def current_player(self):
        return self.players[self.current_player_idx]

    def add(self, player_name):
        if len(self.players) == self.MAX_PLAYERS:
            self.announce(f"Cannot add {player_name}. Game is full (Max {self.MAX_PLAYERS} players).")
            return False

        self.players.append(Player(player_name))
        self.announce(f"{player_name} was added")
        self.announce(f"They are player number {len(self.players)}")
        return True

    def roll(self, roll_value):
        player = self.current_player
        self.announce(f"{player.name} is the current player")
        self.announce(f"They have rolled a {roll_value}")

        if player.in_penalty_box:
            self.handle_penalty_box_turn(player, roll_value)
        else:
            self.handle_normal_turn(player, roll_value)

    def handle_normal_turn(self, player, roll_value):
        player.advance(roll_value, self.BOARD_SIZE)
        self.announce(f"{player.name}'s new location is {player.position}")
        self.announce(f"The category is {self.current_category(player.position)}")
        self.ask_question()

    def handle_penalty_box_turn(self, player, roll_value):
        if roll_value % 2 != 0:
            player.is_getting_out = True
            self.announce(f"{player.name} is getting out of the penalty box")
            self.handle_normal_turn(player, roll_value)
        else:
            self.announce(f"{player.name} is not getting out of the penalty box")
            player.is_getting_out = False

    def ask_question(self):
        category = self.current_category(self.current_player.position)
        self.announce(self.deck.draw_question(category))

    def current_category(self, position):
        if position in [1, 5, 9]: return "Pop"
        if position in [2, 6, 10]: return "Science"
        if position in [3, 7, 11]: return "Sports"
        return "Rock"

    def handle_correct_answer(self):
        player = self.current_player
        if player.in_penalty_box:
            return self.handle_penalty_box_branch(player.is_getting_out)
        else:
            self.add_gold_coin("Answer was correct!!!!")
            still_going = not self.did_player_win()
            self.next_player()
            return still_going

    def wrong_answer(self):
        player = self.current_player
        print("Question was incorrectly answered")
        print(f"{player.name} was sent to the penalty box")
        player.send_to_penalty_box()
        self.next_player()
        return True

    def did_player_win(self):
        return self.current_player.has_won(self.WINNING_COINS)

    def handle_penalty_box_branch(self, is_getting_out):
        if is_getting_out:
            self.add_gold_coin("Answer was correct!!!!")
            still_going = not self.did_player_win()
            self.next_player()
            return still_going
        else:
            self.next_player()
            return True

    def next_player(self):
        self.current_player_idx += 1
        if self.current_player_idx == len(self.players):
            self.current_player_idx = 0

    def add_gold_coin(self, message):
        player = self.current_player
        self.announce(message)
        player.add_coin()
        self.announce(f"{player.name} now has {player.coins} Gold Coins.")

    def announce(self, message):
        print(message)

File: src/main/java/trivia/Game.java
package trivia;

import java.util.ArrayList;
import java.util.LinkedList;



// REFACTOR ME
public class Game implements IGame {
   private static final int BOARD_SIZE = 12;
   private static final int WINNING_COINS = 6;

   ArrayList<Player> players = new ArrayList<>();   

   LinkedList popQuestions = new LinkedList();
   LinkedList scienceQuestions = new LinkedList();
   LinkedList sportsQuestions = new LinkedList();
   LinkedList rockQuestions = new LinkedList();

   int currentPlayer = 0;
   boolean isGettingOutOfPenaltyBox;

   public Game() {
      for (int i = 0; i < 50; i++) {
         popQuestions.addLast("Pop Question " + i);
         scienceQuestions.addLast(("Science Question " + i));
         sportsQuestions.addLast(("Sports Question " + i));
         rockQuestions.addLast(createRockQuestion(i));
      }
   }

   public String createRockQuestion(int index) {
      return "Rock Question " + index;
   }

   public boolean hasEnoughPlayers() {
      return (howManyPlayers() >= 2);
   }

   public boolean add(String playerName) {
      players.add(new Player(playerName));

      System.out.println(playerName + " was added");
      System.out.println("They are player number " + players.size());
      return true;
   }

   public int howManyPlayers() {
      return players.size();
   }

   public void roll(int roll) {
      Player current = players.get(currentPlayer);

      announce(current.name + " is the current player");
      announce("They have rolled a " + roll);

      if (current.inPenaltyBox) {
         if (isOdd(roll)) {
               isGettingOutOfPenaltyBox = true;
               announce(current.name + " is getting out of the penalty box");

               current.advance(roll, BOARD_SIZE);

               announce(current.name + "'s new location is " + current.position);
               announce("The category is " + currentCategory(current.position));
               askQuestion();
         } else {
               announce(current.name + " is not getting out of the penalty box");
               isGettingOutOfPenaltyBox = false;
         }

      } else {
         current.advance(roll, BOARD_SIZE);

         announce(current.name + "'s new location is " + current.position);
         announce("The category is " + currentCategory(current.position));
         askQuestion();
      }
   }

   private void askQuestion() {
      if (currentCategory(players.get(currentPlayer).position) == "Pop")
         announce(popQuestions.removeFirst());
      if (currentCategory(players.get(currentPlayer).position) == "Science")
         announce(scienceQuestions.removeFirst());
      if (currentCategory(players.get(currentPlayer).position) == "Sports")
         announce(sportsQuestions.removeFirst());
      if (currentCategory(players.get(currentPlayer).position) == "Rock")
         announce(rockQuestions.removeFirst());
   }


   // private String currentCategory() {
   //    if (players.get(currentPlayer).position - 1 == 0) return "Pop";
   //    if (players.get(currentPlayer).position - 1 == 4) return "Pop";
   //    if (players.get(currentPlayer).position - 1 == 8) return "Pop";
   //    if (players.get(currentPlayer).position - 1 == 1) return "Science";
   //    if (players.get(currentPlayer).position - 1 == 5) return "Science";
   //    if (players.get(currentPlayer).position - 1 == 9) return "Science";
   //    if (players.get(currentPlayer).position - 1 == 2) return "Sports";
   //    if (players.get(currentPlayer).position - 1 == 6) return "Sports";
   //    if (players.get(currentPlayer).position - 1 == 10) return "Sports";
   //    return "Rock";
   // }

   private String currentCategory(int position) {
      if (position == 1 || position == 5 || position == 9) return "Pop";
      if (position == 2 || position == 6 || position == 10) return "Science";
      if (position == 3 || position == 7 || position == 11) return "Sports";
      return "Rock";
   }

   public boolean handleCorrectAnswer() {
      if (players.get(currentPlayer).inPenaltyBox) {
         return handlePenaltyBoxBranch(isGettingOutOfPenaltyBox);
      } else {
         addGoldCoin("Answer was corrent!!!!"); // <--- Con 'n'
         boolean winner = isGameStillGoing();
         nextPlayer();  
         return winner;
      }
   }

   public boolean wrongAnswer() {
      System.out.println("Question was incorrectly answered");
      System.out.println(players.get(currentPlayer).name + " was sent to the penalty box");
      players.get(currentPlayer).sendToPenaltyBox();

      nextPlayer(); // <--- Mucho más limpio que las 2 líneas de antes
      return true;
   }


   private boolean isGameStillGoing() {
      return !didPlayerWin();
   }

   private boolean isOdd(int roll) {
      return roll % 2 != 0;
   }

   private void announce(Object message) {
      System.out.println(message);
   }

   private boolean didPlayerWin() {
      return players.get(currentPlayer).hasWon(WINNING_COINS);
   }

   private boolean handlePenaltyBoxBranch(boolean isGettingOut) {
      if (isGettingOut) {
         addGoldCoin("Answer was correct!!!!");          
         boolean winner = isGameStillGoing();
         nextPlayer();
         return winner;
      } else {
         nextPlayer();
         return true;
      }
   }

   private void nextPlayer() {
      currentPlayer++;
      if (currentPlayer == players.size()) currentPlayer = 0;
   }

   private void addGoldCoin(String message) {
      announce(message); 
      players.get(currentPlayer).addCoin();
      announce(players.get(currentPlayer).name 
         + " now has " 
         + players.get(currentPlayer).coins 
         + " Gold Coins.");
   }

}

class Player {
   final String name;
   int position = 1; 
   int coins = 0;
   boolean inPenaltyBox = false;

   Player(String name) { this.name = name; }



   public void addCoin() {
      this.coins++;
   }

   public void sendToPenaltyBox() {
      this.inPenaltyBox = true;
   }

   public boolean hasWon(int winningCoins) {
      return this.coins == winningCoins;
   }

   public void advance(int roll, int boardSize) {
      this.position = this.position + roll;
      if (this.position > boardSize) {
         this.position = this.position - 12; 
      }
   }
}